src/resmom/linux/alps.c

```diff
@@ -6390,6 +6677,802 @@ alps_inventory(void)
 	return;
 }
 
+/**
+ *
+ * @brief System Query handling (for KNL Nodes).
+ * 	  Invoked from dep_topology() in mom_mach.c before alps_inventory()
+ * 	  (which handles processing for non-KNL Cray Compute Nodes) is called.
+ *	  Checks if BASIL 1.7 is supported, then makes a System Query request and
+ *	  populates System Query related structures.
+ *
+ * @return void
+ *
+ */
+void
+alps_system_KNL(void)
+{
+	/*
+	 * Determine if ALPS supports the BASIL 1.7 protocol. We are only
+	 * partially supporting the BASIL 1.7 protocol (for the System Query).
+	 */
+	alps_engine_query_KNL();
+
+	if (basil_1_7_supported)
+		log_event(PBSEVENT_DEBUG2, PBS_EVENTCLASS_NODE, LOG_DEBUG, __func__,
+			"This Cray system supports the BASIL 1.7 protocol.");
+	else {
+		log_event(PBSEVENT_DEBUG2, PBS_EVENTCLASS_NODE, LOG_ERR, __func__,
+			"This Cray system does not support the BASIL 1.7 protocol.");
+		return;
+	}
+
+	/*
+	 * Allocate a buffer (requestBuffer_knl) for a new ALPS request (System Query).
+	 * A nonzero return value indicates failure; return at this point without proceeding
+	 * with System Query processing.
+	 */
+	if (new_alps_req_KNL() != 0)
+		return;
+
+	/* Create a System (BASIL 1.7) Query request to fetch KNL information. */
+	snprintf(requestBuffer_knl, UTIL_BUFFER_LEN, "<?xml version=\"1.0\"?>\n"
+		"<" BASIL_ELM_REQUEST " "
+		BASIL_ATR_PROTOCOL "=\"%s\" "
+		BASIL_ATR_METHOD "=\"" BASIL_VAL_QUERY "\" "
+		BASIL_ATR_TYPE "=\"" BASIL_VAL_SYSTEM "\"/>", basilversion_system);
+
+	/*
+	 * This flag is checked in response_start() (a callback function invoked during
+	 * alps_request() processing). Flow of control can arrive at response_start() from
+	 * either alps_system_KNL() or alps_inventory(). This flag helps make this
+	 * distinction.
+	 */
+	is_system_query = 1;
+	/* Issue XML Request. Parse XML Response. Populate appropriate basil.h structs. */
+	brp_knl = alps_request(requestBuffer_knl);
+	is_system_query = 0;
```

Comment:
Can we have alps_request() call inside new_alps_req_KNL()? that will justify to function description.

Reply:
The request are different in different places of the code. I have rename the new_alps_req_KNL to init_KNL_alps_req_buf to make it more meaningful.